jsonschema/tests/test_cli.py

```diff
@@ -173,3 +229,111 @@ def test_piping(self):
+        exit_code = cli.run(
+            {
+                "validator": fake_validator(),
+                "schema": "bad1.json",
+                "instances": ["foo1.json"],
+                "error_format": "{error.message}",
+                "output": "plain",
+            },
+            stdout=stdout,
+            stderr=stderr,
+        )
+        self.assertFalse(stdout.getvalue())
+        self.assertTrue("Failed to parse bad1.json" in stderr.getvalue())
+        self.assertEqual(exit_code, 1)
+
+    def test_instance_parsing_error(self):
+        stdout, stderr = NativeIO(), NativeIO()
+        exit_code = cli.run(
+            {
+                "validator": fake_validator(),
+                "schema": "schema.json",
+                "instances": ["bad1.json", "bad2.json"],
+                "error_format": "{error.message}",
+                "output": "plain",
+            },
+            stdout=stdout,
+            stderr=stderr,
+        )
+        output_err = stderr.getvalue()
+        self.assertFalse(stdout.getvalue())
+        self.assertTrue(
+            "Failed to parse bad1.json" in output_err
+            and "Failed to parse bad2.json" in output_err
+        )
+        self.assertEqual(exit_code, 1)
+
+    def test_stdin_parsing_error(self):
+        stdout, stderr, stdin = NativeIO(), NativeIO(), NativeIO("{foo}")
+        exit_code = cli.run(
+            {
+                "validator": fake_validator(),
+                "schema": "schema.json",
+                "instances": [],
+                "error_format": "{error.message}",
+                "output": "plain",
+            },
+            stdout=stdout,
+            stderr=stderr,
+            stdin=stdin,
+        )
+        self.assertFalse(stdout.getvalue())
+        self.assertTrue("Failed to parse stdin" in stderr.getvalue())
+        self.assertEqual(exit_code, 1)
+
+    def test_pretty_parsing_error(self):
+        stdout, stderr = NativeIO(), NativeIO()
+        exit_code = cli.run(
+            {
+                "validator": fake_validator(),
+                "schema": "bad1.json",
+                "instances": ["foo1.json"],
+                "error_format": "",
+                "output": "pretty",
+            },
+            stdout=stdout,
+            stderr=stderr,
+        )
+        self.assertFalse(stdout.getvalue())
+        self.assertTrue(self.pretty_error_tag in stderr.getvalue())
+        self.assertEqual(exit_code, 1)
+
+    def test_pretty_successful_validation(self):
+        stdout, stderr = NativeIO(), NativeIO()
+        exit_code = cli.run(
+            {
+                "validator": fake_validator(),
+                "schema": "schema.json",
+                "instances": ["foo2.json"],
+                "error_format": "",
+                "output": "pretty",
+            },
+            stdout=stdout,
+            stderr=stderr,
+        )
+        self.assertTrue(self.pretty_success_tag in stdout.getvalue())
```

Comment:
`self.assertIn` will produce slightly nicer error messages.

Reply:
I just discovered `unittest` while writing this, I’m more used to `pytest` where those nice assertions do not exist. Thanks for the tip!